### PR TITLE
Add display title, timestamp and update tool for resource table.

### PR DESCRIPTION
**Pull request by EreMaijala**
Also removes the metadata update from the web-based upgrade tool. 

TODO:
- [x] Should maybe also update record id if a redirected record is found
- [x] Do we need some sort of a notification suggesting to run the command line tool in the web-based upgrade?
- [x] Changelog entry (database changelog; signature changes to ResourcePopulator; note removal of web-based metadata upgrade tool; note addition of CLI tool)
- [x] Add the command to wiki

### CODE REVIEW

File: module/VuFind/src/VuFind/Controller/UpgradeController.php [hunk 1]
```diff
@@ -529,59 +527,6 @@ public function fixduplicatetagsAction()
         return $this->createViewModel();
     }
 
-    /**
-     * Fix missing metadata in the resource table.
-     *
-     * @return mixed
-     * @throws Exception
-     */
-    public function fixmetadataAction()
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/UpgradeController.php:538]: We need to make some theme changes -- at minimum, delete upgrade/fixmetadata.phtml and remove the metadata checkbox from upgrade/getsourceversion.phtml.

File: module/VuFind/src/VuFind/Db/Service/ResourceService.php [hunk 2]
```diff
@@ -109,18 +110,42 @@ public function createEntity(): ResourceEntityInterface
...
+    public function findMetadataToUpdate(?int $lastId, int $limit, ?int $minAge = null, ?string $source = null): array
     {
-        $dql = 'SELECT r '
-            . 'FROM ' . ResourceEntityInterface::class . ' r '
-            . "WHERE r.title = '' OR r.author IS NULL OR r.year IS NULL";
+        $dql = 'SELECT r FROM ' . ResourceEntityInterface::class . ' r';
+        $params = [];
+        if (null !== $minAge) {
+            $date = new DateTime();
+            $date->modify("-$minAge days");
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/ResourceService.php]: Any reason we can't do it like this to save a line (and use a less mutable style)?

```suggestion
            $date = new DateTime("now - $minAge days");
```

File: module/VuFind/src/VuFind/Db/Service/ResourceService.php [hunk 3]
```diff
@@ -109,18 +110,42 @@ public function createEntity(): ResourceEntityInterface
...
 
     /**
-     * Get a set of records that do not have metadata stored in the resource
-     * table.
+     * Get a set of records that are missing metadata in the resource table. If maxAge is specified, this includes also
+     * records that need to be updated.
+     *
+     * @param ?int    $lastId ID of last checked record, or null to start from beginning
+     * @param int     $limit  Limit for results
+     * @param ?int    $minAge Minimum age (in days) for metadata before it needs to be updated
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/ResourceService.php]: Is it worth commenting on what happens when this is null, since the behavior is somewhat non-obvious (though I do understand the reason for it)?

File: module/VuFind/src/VuFind/Db/Service/ResourceService.php [hunk 4]
```diff
@@ -109,18 +110,42 @@ public function createEntity(): ResourceEntityInterface
...
+        if (null !== $minAge) {
+            $date = new DateTime();
+            $date->modify("-$minAge days");
+            $dql .= ' WHERE (r.updated <= :dateThreshold)';
+            $params['dateThreshold'] = $date->format(VUFIND_DATABASE_DATETIME_FORMAT);
+        } else {
+            $dql .= " WHERE (r.title = '' OR r.displayTitle IS NULL OR r.author IS NULL OR r.year IS NULL)";
+        }
+        if ($source) {
+            $dql .= ' AND r.source = :source';
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/ResourceService.php]: Would it be more flexible to use `in (:source)` and cast $source to an array, so we can support `string|string[]` here? We don't need to use that now, but it seems like it would be a simple change that might be handy in the future.

File: module/VuFind/src/VuFind/Db/Service/ResourceService.php [hunk 5]
```diff
@@ -109,18 +110,42 @@ public function createEntity(): ResourceEntityInterface
...
+            $params['source'] = $source;
+        }
+        if (null !== $lastId) {
+            $dql .= ' AND r.id > :lastId';
+            $params['lastId'] = $lastId;
+        }
+        $dql .= ' ORDER BY r.id';
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/ResourceService.php:115]: Is it possible we might want to order by updated date in some cases, to fix the oldest ones first? In a round-robin scenario where we don't update everything at once, I could see that being useful (but of course it would break the "continue using ID number" strategy in play here, so maybe that's not worth changing right now).

File: module/VuFind/src/VuFind/Record/ResourcePopulator.php [hunk 6]
```diff
@@ -167,28 +166,35 @@ public function assignMetadata(ResourceEntityInterface $resource, RecordDriver $
...
             255,
             'UTF-8'
         );
-        if (!empty($author)) {
-            $resource->setAuthor($author);
-        }
+        $resource->setAuthor($author);
 
-        // Try to find a year; if not available, just leave the default null:
+        // Try to find a year; if not available, just set to an empty string:
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Record/ResourcePopulator.php]: If we're using `intval` below, that's not going to yield an empty string. Not sure if the problem here is the code or the comment, but I think something needs adjustment.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommandFactory.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
+<?php
+
+/**
+ * Factory for Util/UpdateResourceMetadata.
+ *
+ * PHP version 8
+ *
+ * Copyright (C) Villanova University 2020.
```
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommandFactory.php]: Might want to tweak the copyright and credits here since this file is all yours!

File: module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+     * @param ResourceServiceInterface $resourceService   Resource service
+     * @param Loader                   $recordLoader      Record loader
+     * @param ResourcePopulator        $resourcePopulator Resource pop
+     */
+    public function __construct(
+        protected ResourceServiceInterface $resourceService,
+        protected Loader $recordLoader,
+        protected ResourcePopulator $resourcePopulator
+    ) {
+        parent::__construct();
```
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php]: Should you pass along a `$name` argument here as in other commands?

File: module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                InputOption::VALUE_REQUIRED,
+                'Number of records to process in a single batch',
+                1000
+            )
+            ->addOption(
+                'min-age',
+                null,
+                InputOption::VALUE_REQUIRED,
+                'Minimum age of a record (in days) before it is refreshed even if metadata is not missing.'
+                . ' By default records with missing metadata are updated, but using this option allows to update all'
```
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php]: Minor grammatical suggestion:

```suggestion
                . ' By default records with missing metadata are updated, but using this option allows updates to all'
```

File: module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+                            . '</comment>'
+                        );
+                        ++$missing;
+                        // Always update the timestamp when running a periodical refresh:
+                        if (null !== $minAge) {
+                            $resource->setUpdated(new DateTime());
+                        }
+                    } else {
+                        $this->resourcePopulator->assignMetadata($resource, $driver);
+                        $resource->setUpdated(new DateTime());
```
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php]: Should the update date adjustment be incorporated into assignMetadata rather than done externally?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, this is off to a great start! See below for a few suggestions.

File: module/VuFind/src/VuFind/Controller/UpgradeController.php (hunk 1 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/UpgradeController.php:538]: I also sort up unwrapped the HTML creation in upgrade/getsourceversion.phtml because the loop no longer saved much.

File: module/VuFind/src/VuFind/Db/Service/ResourceService.php (hunk 4 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Db/Service/ResourceService.php]: I changed it to an array and added support to the command line tool because it was really easy to do while at it.

File: module/VuFind/src/VuFind/Db/Service/ResourceService.php (hunk 5 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Db/Service/ResourceService.php:115]: That'd be nice, but would make iterating so much more complicated that I'm not sure if it's worth it. In a way it can happen also with the current mechanism if you have a suitably high min-age and run the process regularly; if it gets interrupted, the recently updated records won't be checked again on the next run.

File: module/VuFind/src/VuFind/Record/ResourcePopulator.php (hunk 6 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Record/ResourcePopulator.php]: Oops. The database field is an int, so we can't store an empty string to indicate that this value has been set (though a valid year wasn't found). But if we leave it null, the check for missing metadata will find it again and again. Zero is an option, but it's not really right either. I'm not sure if that makes any actual difference, but it would affect sorting of missing values if there are B.C.(E.) years. I ended up using zero anyway for now, but I'm open to other suggestions.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php (hunk 10 above)
Comment (EreMaijala) [reply module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php]: I think it doesn't belong to assignMetadata. We could rename it, but it's also needed when assignMetadata isn't called, so it could cause confusion.

File: module/VuFind/src/VuFind/Db/Service/ResourceService.php [hunk 11]
```diff
@@ -109,18 +110,42 @@ public function createEntity(): ResourceEntityInterface
...
-     * Get a set of records that do not have metadata stored in the resource
-     * table.
+     * Get a set of records that are missing metadata in the resource table. If maxAge is specified, this includes also
+     * records that need to be updated.
+     *
+     * @param ?int  $lastId  ID of last checked record, or null to start from beginning
+     * @param int   $limit   Limit for results
+     * @param ?int  $minAge  Minimum age (in days) for metadata before it needs to be updated, or null to search for
+     * records that are missing metadata
+     * @param array $sources Record source filter
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/ResourceService.php]: You could make this type more specific:


```suggestion
     * @param string[] $sources Record source filter
```

File: module/VuFind/src/VuFind/Db/Service/ResourceServiceInterface.php [hunk 12]
```diff
@@ -64,12 +64,18 @@ public function getResourceById(int $id): ?ResourceEntityInterface;
...
-     * Get a set of records that do not have metadata stored in the resource
-     * table.
+     * Get a set of records that are missing metadata in the resource table. If maxAge is specified, this includes also
+     * records that need to be updated.
+     *
+     * @param ?int  $lastId  ID of last checked record, or null to start from beginning
+     * @param int   $limit   Limit for results
+     * @param ?int  $minAge  Minimum age (in days) for metadata before it needs to be updated, or null to search for
+     * records that are missing metadata
+     * @param array $sources Record source filter
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/ResourceServiceInterface.php]: This should match the above.

```suggestion
     * @param string[] $sources Record source filter
```

(And of course, you'll need to run `phpcbf` if you accept these changes to realign all the other comments.)

File: module/VuFind/src/VuFind/Record/ResourcePopulator.php (hunk 6 above)
Comment (demiankatz) [reply module/VuFind/src/VuFind/Record/ResourcePopulator.php]: I think if we left the value null, sorting might be better -- since null values would sort separately from non-null values, as opposed to 0 which would fall between AD/BCE. Since missing year metadata was only a problem for pre-10.0 releases, maybe the best solution is just to drop null years from the missing metadata check. Really, if we enforce the "upgrade to 10.0 first" rule, the only missing metadata we need to worry about in 11.0 is the display title. It probably doesn't hurt to keep checking other fields, but I don't think leaving year out would be a major issue. If year was really missing, other fields would also be missing.

Beyond that, the only solution I could imagine would be adding another database column like a nullable boolean `year_is_unknown` field. That would work, but it seems like overkill.

File: module/VuFind/src/VuFind/Db/Service/ResourceService.php (hunk 5 above)
Comment (demiankatz) [reply module/VuFind/src/VuFind/Db/Service/ResourceService.php:115]: Fair enough, I agree that it's not worth the extra complexity.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php (hunk 10 above)
Comment (demiankatz) [reply module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php]: Fair enough!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, great progress -- see below for the latest comments.

File: module/VuFind/src/VuFind/Record/ResourcePopulator.php (hunk 6 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Record/ResourcePopulator.php]: I thought about it too, but an additional column just for that feels overkill for me too. Now it's left null and removed from the query for missing metadata.

File: module/VuFind/src/VuFind/Record/ResourcePopulator.php [hunk 13]
```diff
@@ -167,40 +166,46 @@ public function assignMetadata(ResourceEntityInterface $resource, RecordDriver $
...
         );
-        if (!empty($author)) {
-            $resource->setAuthor($author);
-        }
+        $resource->setAuthor($author);
 
         // Try to find a year; if not available, just leave the default null:
         $dates = $driver->tryMethod('getPublicationDates');
-        if (isset($dates[0]) && strlen($dates[0]) > 4) {
+        $year = $dates[0] ?? null;
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Record/ResourcePopulator.php]: While we're refactoring this anyway, I wonder about a couple more potentially helpful changes:

1.) Why are we only looking at the first date? Why not loop through the array and keep trying until we find something we can use?

2.) Why are we using the date converter's `convertFromDisplayDate`? There's no reason to assume that metadata is going to conform to our configured display format. Maybe it would be wiser to just use a generic `DateTime` object to extract the year.

What do you think?

**Review by demiankatz — CHANGES_REQUESTED** [review]
I ran out of time before I could test this today, but I did think of some changes that might be worth making before I do my final round of hands-on testing. What do you think?

File: module/VuFind/src/VuFind/Record/ResourcePopulator.php (hunk 13 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Record/ResourcePopulator.php]: I was wondering about that but dared not touch anything. :) I'll revise.

Comment (EreMaijala) [reply module/VuFind/src/VuFind/Record/ResourcePopulator.php]: Now improved. Wasn't that simple because something like '2025' is parsed as 20:25 today, so I added a bit more logic, and tests.

File: module/VuFind/src/VuFind/Controller/UpgradeController.php (hunk 1 above)
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/UpgradeController.php:538]: Regarding your checkbox about some kind of message in the web-based upgrade tool about the new command line tool, maybe what we should do is keep fixmetadataAction, but change it so that all it does is do a minimal lookup for potential problems, and if it finds any, add a warning message instructing the user to run the command line tool to fix things. If we do this, I think it's fine to leave the "metadata update" skip checkbox out, since we've reduced this from a data-changing and expensive activity to a simple validation check.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+        $lastId = null;
+        $output->writeln('<info>Updating resource metadata</info>');
+        while ($resources = $this->resourceService->findMetadataToUpdate($lastId, $batch, $minAge, $backend)) {
+            $recordIds = [];
+            array_walk(
+                $resources,
+                function ($resource) use (&$recordIds) {
+                    $recordIds[] = ['id' => $resource->getRecordId(), 'source' => $resource->getSource()];
+                }
+            );
```
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php]: Sorry if I'm missing some subtlety that makes this a bad idea, but why not simplify to this?


```suggestion
            $recordIds = array_map(
                fn ($resource) => ['id' => $resource->getRecordId(), 'source' => $resource->getSource()],
                $resources
            );
```

File: module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+                        $resource->setRecordId($driverRecordId);
+                        ++$redirected;
+                    }
+                    ++$updated;
+                }
+            }
+            $this->persistenceManager->flushEntities();
+            foreach ($resources as $resource) {
+                $this->persistenceManager->detachEntity($resource);
+            }
```
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/UpdateResourceMetadataCommand.php]: Could we simplify?


```suggestion
            array_walk($resources, [$this->persistenceManager, 'detachEntity']);
```

File: module/VuFind/src/VuFind/Record/ResourcePopulator.php [hunk 16]
```diff
@@ -194,20 +187,21 @@ public function assignMetadata(ResourceEntityInterface $resource, RecordDriver $
...
-            try {
-                $year = $this->dateConverter->convertFromDisplayDate('Y', $dates[0]);
-            } catch (DateException) {
-                // If conversion fails, don't store a date
-                $year = null;
+        // Try to find a year; if not available, just set to null:
+        $year = null;
+        foreach ($driver->tryMethod('getPublicationDates', [], []) as $pubDate) {
+            // Try to extract a year from a string like '2025', 'Ⓟ2025' or 'copyright 2025':
+            if (preg_match('/^[^\d]*?(-?\d+)$/', $pubDate, $matches)) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Record/ResourcePopulator.php]: Is there a risk that something like `copyright 1993-2025` will resolve as `-2025` based on this regular expression, or am I misreading something?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala -- see below for a few more comments.

File: module/VuFind/src/VuFind/Record/ResourcePopulator.php (hunk 16 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Record/ResourcePopulator.php]: No, because the leading content cannot include numbers. But it didn't find `1993` either, so I relaxed the rule a bit to allow for trailing content as well.

Comment (EreMaijala) [reply module/VuFind/src/VuFind/Record/ResourcePopulator.php]: Then I added a bit smarter year range parsing to make it take the end year instead of the start year.

File: module/VuFind/src/VuFind/Controller/UpgradeController.php (hunk 1 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/UpgradeController.php:538]: I restored the action as an information-only version. And cleaned up translation strings that are no longer needed.

File: themes/bootstrap5/templates/upgrade/fixmetadata.phtml [hunk 17]
```diff
@@ -6,21 +6,13 @@
...
-<p>Some of the items in your resource table appear to be missing metadata.  Adding this metadata may take some time,
-but it will improve the user experience by allowing proper sorting of favorites and tagged records.</p>
-
-<br>
+<p>
+  Some of the items in your resource table appear to be missing metadata. It is recommended to run the
+  <strong>util/update_resource_metadata</strong> utility to update the resources. It can be a time-consuming process,
+  but it will improve the user experience by allowing proper sorting of favorites and tagged records, as well as
+  displaying basic metadata for records that are not available in the search index.
+</p>
```
Comment (demiankatz) [inline themes/bootstrap5/templates/upgrade/fixmetadata.phtml:14]: What do you think about adding something like this?

```suggestion
</p>
<p>
  See the <a href="https://vufind.org/wiki/administration:command_line_utilities">Command Line Utilities</a> wiki page for more details.
</p>
```

File: module/VuFind/tests/unit-tests/src/VuFindTest/Record/ResourcePopulatorTest.php [hunk 18]
```diff
@@ -47,22 +46,51 @@
...
+            [['-100'], -100],
+            [['0'], 0],
+            [['foo', '1999'], 1999],
+            [['©1999'], 1999],
+            [['Ⓟ1999'], 1999],
+            [['copyright 1999'], 1999],
+            [['2020-2025'], 2025],
+            [['2020 - 2025'], 2025],
+            [['copyright 2020-2025'], 2025],
+            [['copyright 2020 -2025'], 2025],
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Record/ResourcePopulatorTest.php:69]: Do you think it might also be worth supporting edge cases where years are uncertain? For example, something like `1900?-1925?`. I imagine adding an optional `\??` after the `\d{4}` blocks in your first regex might cover this without conflicting with other scenarios... but if you think this is overkill, I won't insist!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, I think we're nearly done -- see below for a couple of small suggestions.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Record/ResourcePopulatorTest.php (hunk 18 above)
Comment (EreMaijala) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Record/ResourcePopulatorTest.php:69]: Done! I suppose this could still keep on giving, but at least it's now so much better than before. :)

**Review by demiankatz — CHANGES_REQUESTED** [review]
I ran into an interesting problem here which is not directly related to these changes, but which is revealed by them... so I'm not sure if I should approve and merge this and solve it separately, or try to address it here. Either way, I'm short on time today, so I thought I'd leave this comment, and we can discuss further tomorrow.

Here's what happened:

1.) I wanted to test the upgrade process, so I started a test instance on the dev branch, created an account, and saved some favorites.

2.) Next I switched to this PR's branch and ran the upgrade process.

3.) After upgrading the database, I immediately encountered a fatal error: the query to check for missing metadata was using a field that Doctrine didn't recognize, because it had cached the database definitions from prior to the database upgrade.

Clearly we need to revise the upgrade process to clear the cache to prevent this type of problem. The question that I unfortunately don't have time to research right now is whether there's a call we can make on some piece of existing Doctrine code to accomplish that, or if we just need to brute-force clear out the file cache set up [in module.config.php](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/config/module.config.php#L657-L663).

Let me know if you have any thoughts on how best to proceed, and I'll be happy to put some more time/effort into this tomorrow when my schedule opens up again.

**Comment by EreMaijala** [comment]
@demiankatz Oh, yes, that's an issue. I think I know a solution. I'll test and add it here if it works.

**Comment by EreMaijala** [comment]
@demiankatz Should work now. :)

**Comment by EreMaijala** [comment]
@demiankatz And now it's sprinkled to a couple of strategic places in the code.

**Comment by EreMaijala** [comment]
Now also changed to in-memory caching in CLI mode.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, this does seem to be working well for me now.

However, before I merge, a couple more thoughts/concerns:

1.) I wonder if we need to put clearDoctrineMetadataCache somewhere better than the UpgradeController. It works fine there in the short term, but I wonder if we need to put it in a service so it's more reusable and the dependency injection is more clear. I'm not sure _which_ service would be appropriate though; maybe it's something new, in which case I guess we can wait until we have a use case.

2.) I remain concerned that there are some situations where running the command line database upgrade will break the running site because the file cache won't get cleared. Due to file permissions, there's probably no straightforward solution to this. I guess a couple of options would include having the database upgrade tool TRY to clear the cache and ask the user to do it if it can't do it itself, OR to have it remind the user to run through the web-based upgrade just to ensure the cache gets cleared. Neither of these is great, but I can't think of anything better. Or am I worrying about an unlikely scenario?

**Comment by EreMaijala** [comment]
@demiankatz 

1. I had this great idea of adding VuFind\Db\EntityManager that could include the method for clearing the cache, but didn't notice that Doctrine's EntityManager is final. So that's a no-go. I decided to reduce it to a one-liner, so if the cache needs to be cleared anywhere else, the dependency is EntityManager and this single line call.
2. I added a message to the command line tool. If we dump the SQL, clearing the cache could be premature, and due to possible permissions issues I thought it would be best to leave it to the user running the upgrade.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala, everything seems to be working as expected now!
